src/protocol.rs

```diff
@@ -1,42 +1,553 @@
+        }
+    }
+}
+
+impl AddressSegment for IPFSSegment {
+    const STREAM_LENGTH: usize = 34;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::IPFS
+    }
+
+    fn from_stream(stream: &mut io::Read) -> Result<Self> {
+        //FIXME: `cid::Prefix` creates its own `Cursor` instead of
+        //       overloading `io::Read`
+
+        // Read CID hash from stream
+        let cid = Self::_read_cid_polyfill(stream).map_err(|err| {
+            println!("CID Parsing failed!");
+            err
+        })?;
+
+        Ok(IPFSSegment(cid))
+    }
+
+    fn to_stream(&self, mut stream: &mut io::Write) -> io::Result<()> {
+        let bytes = self.0.to_bytes();
+
+        stream.write_varint(bytes.len())?;
+        stream.write_all(bytes.as_ref())?;
+
+        Ok(())
+    }
+}
+
+impl FromStr for IPFSSegment {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        //FIXME: `cid::Cid` does not implement `FromStr`
+        Ok(IPFSSegment(cid::Cid::from(s)?))
+    }
+}
+
+impl hash::Hash for IPFSSegment {
+    fn hash<H>(&self, state: &mut H) where H: hash::Hasher {
+        //FIXME: `cid::Cid` does not derive `Hash`
+        state.write_usize(self.0.version as usize);
+        state.write_usize(self.0.codec as usize);
+        hash::Hash::hash(&self.0.hash, state);
+    }
+}
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct HTTPSegment;
+
+derive_for_empty_segment!(HTTPSegment: Display, FromStr, AddressSegment<HTTP>);
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct HTTPSSegment;
+
+derive_for_empty_segment!(HTTPSSegment: Display, FromStr, AddressSegment<HTTPS>);
+
+
+
+//TODO: Properly parse and format union addresses
+pub struct OnionSegment(pub Box<[u8; 80]>);
+
+derive_for_empty_segment!(OnionSegment: Display);
+impl AddressSegment for OnionSegment {
+    const STREAM_LENGTH: usize = 80;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::Onion
+    }
+
+    fn from_stream(_: &mut io::Read) -> Result<Self> {
+        Err(Error::ParsingError)
+    }
+
+    fn to_stream(&self, _: &mut io::Write) -> io::Result<()> {
+        Err(io::Error::from(io::ErrorKind::BrokenPipe))
+    }
+}
+
+impl FromStr for OnionSegment {
+    type Err = Error;
+
+    fn from_str(_: &str) -> Result<Self> {
+        Err(Error::ParsingError)
+    }
+}
+
+// Manual implementation of the default derives because Rust arrays
+// currently only implement them for up to 32 entries
+impl fmt::Debug for OnionSegment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // Workaround for the fact that an `&[u8]` cannot currently be cast
+        // to an `&fmt::Debug`
+        struct F<'a>(&'a [u8]);
+        impl<'a> fmt::Debug for F<'a> {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                fmt::Debug::fmt(&self.0, f)
+            }
+        }
+        let debug_ref = F(&*self.0 as &[u8]);
+
+        f.debug_tuple("OnionSegment").field(&debug_ref).finish()
```

Comment:
I didn't know that `&[u8]` does not implement `Debug` 😞  In my projects, we usually display it as [hex](https://docs.rs/rustc-hex/1.0.0/rustc_hex/trait.ToHex.html#tymethod.to_hex).

Reply:
It does implement `Debug`, but one cannot cast `&[u8]` (an unsized type) to `&Debug` apparently. Hence the (not too hard to understand IMHO) workaround.